### DISCUSSION TITLE
Validation on FIELD ?

### DISCUSSION THREAD

**Discussion by aoudiamoncef** [graphql-java/graphql-java#2574] [Q&A]
Hi @bbakerman, @andimarek,

I would like to know if it's possible to implement a validation on FIELD with ` GraphQL Java 17.3` ? 

It's the same use case as mentioned in [QueryDirectivesIntegrationTest ](https://github.com/graphql-java/graphql-java/blob/ade69c09aa63418490231650b84bbe261adfeb24/src/test/groovy/graphql/execution/directives/QueryDirectivesIntegrationTest.groovy).

I saw those files: 
[Introspection](https://github.com/graphql-java/graphql-java/blob/ade69c09aa63418490231650b84bbe261adfeb24/src/main/java/graphql/introspection/Introspection.java#L409)
[Directives](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/Directives.java) 
[DirectiveTest](https://github.com/graphql-java/graphql-java/blob/master/src/test/groovy/graphql/language/DirectiveTest.groovy)
[IntrospectionWithDirectivesSupportTest](https://github.com/graphql-java/graphql-java/blob/master/src/test/java/graphql/introspection/IntrospectionWithDirectivesSupportTest.groovy)

Thanks :)
Moncef AOUDIA

**Comment by dugenkui03** [comment]
@aoudiamoncef Can you please give more detail? what exactly did you expected the capability of validation directive? like @timeout. Maybe I can offer some way to do this.

**Reply by aoudiamoncef** [reply]
Hi, 

I'm trying to have client side directive like @skip 

Client could add @uppercase on its query field for example.

**Reply by dugenkui03** [reply]
You can acquire the **query directive info** and the DataFetcher  object in `Instrumentation#instrumentDataFetcher`, and alter the behavior of `DataFetcher#get`.

That's my way, you can transform the value by[`@map`](https://github.com/graphql-calculator/graphql-calculator/blob/main/README_EN.md#feature-showcase).

That's the implementation of Schema directive.
https://github.com/graphql-java/graphql-java/blob/e0eb64c17fd7a15258844a936cdbdd3a99317344/src/test/groovy/graphql/schema/idl/SchemaGeneratorDirectiveHelperTest.groovy#L297

**Comment by bbakerman** [comment]
There is a couple of ways to inject behavior into a field.  As @dugenkui03 says you can use `Instrumentation#instrumentDataFetcher` to change the data fetcher per request and it therefor you can adjust the fields fetching behavior.

The other is to change the data fetcher allotted to the field ONCE at schema build time.

The callback mechanism `SchemaDirectiveWiring` allows you to change a schema element (beyond what was built via the SDL text.

This gives you the directives on the element (type / field / arg etc..) and you can change things.  In the case of fields you can also change the data fetcher that has been associated with the field.

The final way is much like `SchemaDirectiveWiring` and thats `graphql.schema.SchemaTransformer` - this is a mechanism to also change schema structure but you get a variable of type ` GraphQLCodeRegistry.Builder.class` in the context and this can be used to get and change data fetchers

All of this talk of data fetchers is about how they are the way in graphql-java to inject code per field.

You might be interested in https://github.com/graphql-java/graphql-java-extended-validation as well

**Reply by aoudiamoncef** [reply]
Thank you for detailed explanations 😇

I'm already using GraphQL Java extended validation in my project.
